components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar_impl.hpp

```diff
@@ -0,0 +1,47 @@
+#ifndef SHOC_CALC_SHOC_VARORCOVAR_IMPL_HPP
+#define SHOC_CALC_SHOC_VARORCOVAR_IMPL_HPP
+
+#include "shoc_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace shoc {
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::calc_shoc_varorcovar(
+  const MemberType&            team,
+  const Int&                   nlev, 
+  const Scalar& tunefac,
+  const uview_1d<const Spack>& isotropy_zi, 
+  const uview_1d<const Spack>& tkh_zi,
+  const uview_1d<const Spack>& dz_zi,
+  const uview_1d<const Spack>& invar1, 
+  const uview_1d<const Spack>& invar2, 
+  const uview_1d<Spack>&       varorcovar)
+{
+  const Int nlev_pack = ekat::pack::npack<Spack>(nlev);
+  const auto sinvar1 = scalarize(invar1);
+  const auto sinvar2 = scalarize(invar2);
+
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team, nlev_pack), [&] (const Int& k) {
+    // Calculate shift
+    Spack up_grid1, grid1, up_grid2, grid2;
+    auto range_pack1 = ekat::pack::range<IntSmallPack>(k*Spack::n);
+    auto range_pack2 = range_pack1;
+    range_pack2.set(range_pack1 < 1, 1); // don't want the shift to go below zero. we mask out that result anyway
+    ekat::pack::index_and_shift<-1>(sinvar1, range_pack2, grid1, up_grid1);
+    ekat::pack::index_and_shift<-1>(sinvar2, range_pack2, grid2, up_grid2);
+
+    const Spack grid_dz = 1 / dz_zi(k);
+    const Spack grid_dz2 = ekat::pack::square(grid_dz); // vertical grid diff squared
+
+    // Compute the variance or covariance
+    varorcovar(k).set(range_pack1 > 0 && range_pack1 < nlev, tunefac*isotropy_zi(k)*tkh_zi(k)*grid_dz2*(up_grid1 - grid1)*(up_grid2 - grid2));
```

Comment:
Small naming quibble: `(up_)grid1,2` sound like they are variables describing the grid, but in fact they are `invar1,2` values. So you might name them something like `invar1_s`, `invar1_sm1`, where `s` is the conventional index for the pack slot and `sm1` is for `s-1`.

Reply:
I agree with your suggestion. Names have been changed.